### PR TITLE
04 basic struct

### PR DESCRIPTION
**Pull request by kryvosheiaivan**

### CODE REVIEW

File: 04_basic_struct/lkm_example.c [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+ssize_t procamp_store(struct kobject *kobj, struct kobj_attribute *attr,
+		const char *buf, size_t count)
+{
+	int len = strlen(buf);
+
+	if (len <= 0)
+		return -1;
+	pr_debug("storing %s len %d\n", buf, len);
+
+	struct my_obj *obj = vmalloc(1 * sizeof(struct my_obj));
```
Comment (Anton-Soroka) [inline 04_basic_struct/lkm_example.c]: vmalloc is not best option here
generally better to use kmalloc if physically contiguous memory is not required

File: 03_module/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+## simple module home work:  create the simplest kernel loadable module
+
+1. Create a simple kernel module (C source and makefile).
+2. Add some debug prints.
+3. Check your source with the checkpatch.pl
+4. Add module parameter and implement different module_init return codes 
+   dependent on this parameter (OK/Error).
+5. Get kernel logs for both cases.
+Please add your sources and makefile into folder '03_module'
```
Comment (Anton-Soroka) [inline 03_module/README.md]: remove this file 
it is not needed here

File: hwdetect/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+## bash home work:  create hot plugged hardware detector
+Create the script 'hwdetect.sh' which detects connected hardware.
+For monitor hardware from user level please use /dev folder,
+output from lsusb and i2detect (optional).
+Most interesting devices are:
+ - USB to TTL convertors,
+ - flash drives,
+ - SD cards,
+ - i2c devices.
+Please add your scripts in folder 'hwdetect'
```
Comment (Anton-Soroka) [inline hwdetect/README.md]: remove this file 
it is not needed here

File: scissors/README.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+The game should be simple, the goal is to create a git project with proper commit order and structure. E.g.:
+```
+> Please choose: rock (r) - paper (p) - scissors (s)
+> r
+> You choose rock, I choose paper
+> I win: paper beats rock
+```
+Please add your sources into folder 'scissors'
+
+Optional task: add printing md5 checksum of computer's choice before human's move.
```
Comment (Anton-Soroka) [inline scissors/README.md]: remove this file 
it is not needed here

File: 04_basic_struct/lkm_example.c [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	kobj_inst = kobject_create_and_add("gl_procamp", kernel_kobj);
+
+	if (!kobj_inst)
+		return -1;
+
+	rc = sysfs_create_file(kobj_inst, &procamp_attr.attr);
+
+	if (rc) {
+		pr_debug("failed to create the foo file in /sys/kernel/kobject_example\n");
+		return -1;
```
Comment (Anton-Soroka) [inline 04_basic_struct/lkm_example.c:61]: need to remove kobject before this return

File: 04_basic_struct/lkm_example.c [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	return rc;
+}
+
+void deinit_sysfs_attr(void)
+{
+	if (kobj_inst)
+		kobject_put(kobj_inst);
```
Comment (Anton-Soroka) [inline 04_basic_struct/lkm_example.c:72]: need to call sysfs_remove_file before kobject_put

**Review by Anton-Soroka — CHANGES_REQUESTED** [review]
Changes required:
1. fix compilation errors and warnings
image

2. make changes according to comments
3. remove extra files by removing extra commits using interactive rebase

Ivan. You have errors because you compiled for ARM while I was compiling for my PC,
I could not launch QEMU for module 4. In 5 I find out the problem with QEMU and fixed it.
So I changed  Makefile and now youll be able to compile.

File: 04_basic_struct/lkm_example.c (hunk 6 above)
Comment (kryvosheiaivan) [reply 04_basic_struct/lkm_example.c:72]: thanks for that

File: 04_basic_struct/lkm_example.c (hunk 5 above)
Comment (kryvosheiaivan) [reply 04_basic_struct/lkm_example.c:61]: right

File: 04_basic_struct/lkm_example.c (hunk 1 above)
Comment (kryvosheiaivan) [reply 04_basic_struct/lkm_example.c]: I had problems with that, kmalloc solves it!

**Review by Anton-Soroka — APPROVED** [review]
Looks OK!
Please merge!

**Pull request merged**